File: prototype/rails/src/test/actions/favorites/toggle_favorite_test.rb
require "test_helper"

class ToggleFavoriteTest < ActiveSupport::TestCase
  FavoriteChange = Domain::Shop::FavoriteChange

  setup do
    @toggle = Favorites::ToggleFavorite.new
    @shopper = create_verified_customer
    @listing = create_listing
  end

  test "it saves a favorite and records the event" do
    change = @toggle.call(customer: @shopper, listing: @listing, now: moment("2026-08-20 09:00:00"))

    assert_equal FavoriteChange::ADDED, change
    assert @shopper.favorites.exists?(listing: @listing)
    assert_equal ["favorite"], @listing.events.pluck(:event_type)
  end

  test "toggling twice drops the favorite and records the event" do
    @toggle.call(customer: @shopper, listing: @listing, now: moment("2026-08-20 09:00:00"))

    change = @toggle.call(customer: @shopper, listing: @listing, now: moment("2026-08-20 09:05:00"))

    assert_equal FavoriteChange::REMOVED, change
    refute @shopper.favorites.exists?(listing: @listing)
    assert_equal %w[favorite unfavorite], @listing.events.order(:occurred_at).pluck(:event_type)
  end

  test "it records the event against the visitor who saved it" do
    @toggle.call(customer: @shopper, listing: @listing, now: moment("2026-08-20 09:00:00"))

    assert_equal @shopper.id, @listing.events.sole.customer_id
  end

  test "one visitor saving leaves another visitor alone" do
    other = create_verified_customer
    @toggle.call(customer: @shopper, listing: @listing, now: moment("2026-08-20 09:00:00"))

    assert_equal FavoriteChange::ADDED,
      @toggle.call(customer: other, listing: @listing, now: moment("2026-08-20 09:01:00"))
    assert @shopper.favorites.exists?(listing: @listing)
  end
end

File: prototype/node/src/app/actions/favorites/toggle-favorite.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import type { CustomerId, ListingId } from '../../core/ids/entity-ids.ts'
import { toggleFavorite } from './toggle-favorite.ts'
import type { AppDatabase } from '../../db/database.ts'
import { createCustomer, createListing, createSeller, openCommerceWorld } from '../../test/commerce-world.ts'

test('it saves a favorite and records a favorite event', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shopperId = await createCustomer(context)
  const sellerId = await createSeller(context)
  const art = await createListing(context, sellerId)

  const change = await toggleFavorite(context, { customerId: shopperId, listingId: art.id })

  assert.equal(change, 'added')
  assert.ok(await isFavorited(world.db, shopperId, art.id))
  assert.deepEqual(await eventTypes(world.db, art.id), ['favorite'])
})

test('toggling twice drops the favorite and records an unfavorite event', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shopperId = await createCustomer(context)
  const sellerId = await createSeller(context)
  const art = await createListing(context, sellerId)
  await toggleFavorite(context, { customerId: shopperId, listingId: art.id })

  const change = await toggleFavorite(context, { customerId: shopperId, listingId: art.id })

  assert.equal(change, 'removed')
  assert.equal(await isFavorited(world.db, shopperId, art.id), false)
  assert.deepEqual(await eventTypes(world.db, art.id), ['favorite', 'unfavorite'])
})

test('the event is recorded against the visitor who saved it', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shopperId = await createCustomer(context)
  const sellerId = await createSeller(context)
  const art = await createListing(context, sellerId)

  await toggleFavorite(context, { customerId: shopperId, listingId: art.id })

  const event = await world.db
    .selectFrom('listingEvents')
    .select('customerId')
    .where('listingId', '=', art.id)
    .executeTakeFirstOrThrow()

  assert.equal(event.customerId, shopperId)
})

test("one visitor saving leaves another visitor's favorites alone", async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shopperId = await createCustomer(context)
  const otherId = await createCustomer(context)
  const sellerId = await createSeller(context)
  const art = await createListing(context, sellerId)
  await toggleFavorite(context, { customerId: shopperId, listingId: art.id })

  const change = await toggleFavorite(context, { customerId: otherId, listingId: art.id })

  assert.equal(change, 'added')
  assert.ok(await isFavorited(world.db, shopperId, art.id))
})

async function isFavorited(db: AppDatabase, customerId: CustomerId, listingId: ListingId): Promise<boolean> {
  const row = await db
    .selectFrom('favorites')
    .select('id')
    .where('customerId', '=', customerId)
    .where('listingId', '=', listingId)
    .executeTakeFirst()

  return row !== undefined
}

async function eventTypes(db: AppDatabase, listingId: ListingId): Promise<string[]> {
  const rows = await db
    .selectFrom('listingEvents')
    .select('eventType')
    .where('listingId', '=', listingId)
    .orderBy('occurredAt')
    .orderBy('id')
    .execute()

  return rows.map((row) => row.eventType)
}
